Add GitHub Copilot instructions (#1698)
PrivateBin/PrivateBin: .github/copilot-instructions.md

## Repository Overview

**PrivateBin** is a minimalist, open-source online pastebin where the server has zero knowledge of the pasted data. All data is encrypted/decrypted in the browser via 256-bit AES (Galois Counter Mode). The project is a refactored fork of ZeroBin focused on extensibility and additional features.

- **Main Use Case:** Secure, ephemeral sharing of text/code, with encryption happening exclusively client-side.
- **Languages:** PHP (~48%), JavaScript (~35%), CSS (~17%), with some legacy browser support (see `legacy.js`).
- **Type:** Web application (pastebin), with both server (PHP) and client (JavaScript) components.

## Build & Validation Instructions

### Prerequisites

- **PHP:** 7.4+ (recommended: latest stable 7.x or 8.x)
- **Composer:** For dependency management (`composer install`)
- **Node.js & npm:** Only required for running JavaScript unit tests. **Main JS logic must remain browser-compatible!**
- **Recommended Environment:** Unix-like OS (Linux, macOS) with Apache or Nginx for full demo.

### Bootstrap & Install

1. **Clone the repository** and enter its root directory.
2. **Install PHP dependencies:**
   ```sh

>    composer install

Reviewer: Not sure if this would cause issues or not: If these instructions are applied to a test environment, this is fine, but this should not be run on a checked out repo where development occurs in, because then one might accidentally commit dev dependencies or an unoptimized autoloader. In our makefile we give this on the composer target: `composer update --no-dev --optimize-autoloader` (used to update production dependencies, does not include the GPC and AWS libraries)

Reviewer: Hm, I wonder if a solution would be to install phpunit, all our requirements and the google cloudstorage library `--global` which installs them into the user home (not globally for the whole system, which would require root privs). That way it is not installed into the projects vendor directory, which should make it possible to run the unit tests and do development in the same environment.

Author: Also updated this for the DevContainer then: https://github.com/PrivateBin/PrivateBin/pull/1709

Reviewer: Sorry, I'm misleading you with the update one, that is for a different purpose (updating the dependencies in the repo). To install the libraries and phpunit I'd now suggest:

```suggestion
   composer require --global google/cloud-storage phpunit/phpunit
```

   ```
   - Always run this before building or testing PHP code.
   - If you receive permission errors, verify `vendor/` is writable.

3. **Install JS dependencies (for test only):**
   ```sh
   cd ./js
   npm install
   ```
   - Only required for running JS tests. Not needed for building or running the app.

### Build

- **No explicit build step** for PHP. The web app is served directly from source.
  - This means **composer directory** need to be comitted (_except_ of big optional dependences like Google Cloud like GCS support or similar!)
- **For JavaScript:** There is no webpack/bundler step for release; browser JS is written in compatible ES5+ syntax except in `legacy.js` (ES3).

### Run

- **PHP Server Mode:** Use Apache/Nginx with PHP, pointing the web root to the repo root or the `public/` directory (if present).

> - **Demo:** Open `index.php` in a browser (via your web server).

Reviewer: ```suggestion
- **Demo:** Open the root directory served by the web server in a browser. This should call the index.php in the repositories root directory.
```

  ```sh
  npm run lint
  ```
  - Check for configuration in `.eslintrc.*` files.

### Validation / CI

- **GitHub Actions:** CI runs `composer install`, `phpunit`, and JS tests on PRs and pushes.
- **Pre-commit:** Always run both PHP and JS tests before submitting PRs. Fix any warnings or errors.

## Project Layout & Structure

- **Root files:**
  - `README.md`: Project overview ([view full](../README.md)).
  - `composer.json`, `composer.lock`: PHP dependencies.
  - `package.json`: JS test/lint dependencies (not for production JS).
  - `.github/workflows/`: CI configuration.
  - `cfg/`: Default configuration files.
  - `js/`: Main client logic (browser JS), including:
    - `legacy.js`: Must remain compatible with legacy browsers (ES3). **Do not use modern JS here.**
    - `privatebin.js`: Core encryption and paste interface logic.
  - `tpl/`: HTML templates.
  - `css/`: Stylesheets.

- **Testing & Validation:**
  - `tst/`: Contains PHP unit tests.
  - `js/test`: Contains JS unit tests.
  - `phpunit.xml`: PHPUnit config.
  - JS test files may use Node.js features; browser JS must not.

  - **Encryption:** Only client-side in JS with WebCrypto.

>   - **Backend:** Serves encrypted blobs, never sees plaintext.

Reviewer: ```suggestion
  - **Backend:** Serves encrypted blobs (as base64 encoded strings) and plaintext meta data in JSON format. APIs are designed for WORM (write once, read many) usage. Once stored content is never updated, only deleted, if delete token is sent, has expired as per meta data or immediately upon reading for the first time, if meta data was set to burn-after-reading.
```

>   - **Legacy Support:** `js/legacy.js` must remain ES3 for feature detection in old browsers.

Reviewer: ```suggestion
  - **Legacy Support:** `js/legacy.js` must remain compatible with IE4 and Netscape for feature detection of ancient browsers.
```

  - **Configuration:** See `cfg/conf.php` and [wiki](https://github.com/PrivateBin/PrivateBin/wiki/Configuration) for available options.

## Automated Checks

> - **GitHub CI:** On PRs, runs `composer install`, `phpunit`, and JS tests.

Reviewer: Isn't this section redundant to the above "Validation / CI" one? Do they serve a different purpose of can they be unified?

Author: Indeed, LLMs are still just word machines 🙃 😅 

```suggestion
```
